src/application/action/main.ts

```diff
@@ -0,0 +1,139 @@
+import fs from 'node:fs';
+import os from 'node:os';
+
+import * as core from '@actions/core';
+import * as exec from '@actions/exec';
+import path from 'path';
+
+interface Inputs {
+  githubRepo: string;
+  localRegistry: string;
+  metadataTemplate: string;
+  moduleVersion: string;
+  patch: string;
+  presubmit: string;
+  sourceTemplate: string;
+  tag: string;
+  templatesDir: string;
+}
+
+async function main() {
+  try {
+    const inputs: Inputs = {
+      // GitHub doesn't actually validate inputs that have the `required`
+      // property so we need to verify that ourselves with { required: true }.
+      // https://github.com/actions/runner/issues/1070
+      githubRepo: core.getInput('github-repository'),
+      localRegistry: core.getInput('local-registry', { required: true }),
+      metadataTemplate: core.getInput('metadata-template'),
+      moduleVersion: core.getInput('module-version', { required: true }),
+      patch: core.getInput('patch'),
+      presubmit: core.getInput('presubmit'),
+      sourceTemplate: core.getInput('source-template'),
+      tag: core.getInput('tag'),
+      templatesDir: core.getInput('templates-dir'),
+    };
+
+    const templatesDir = validateAndCreateTemplatesDir(inputs);
+
+    const cliBin = getCliBin();
+    const cliArgs = [
+      'create-entry',
+      `--templates-dir=${templatesDir}`,
+      `--local-registry=${inputs.localRegistry}`,
+      `--module-version=${inputs.moduleVersion}`,
+    ];
+
+    if (inputs.githubRepo) {
+      cliArgs.push(`--github-repository=${inputs.githubRepo}`);
+    }
+    if (inputs.tag) {
+      cliArgs.push(`--tag=${inputs.tag}`);
+    }
+
+    const code = await exec.exec('node', [cliBin, ...cliArgs]);
```

Comment:
curious, as we're already in a node process, why not let esbuild include the code needed for the action to re-use whatever functionality we want? and have CLI and action just be alternative entry points to that shared functionality?

Having a subprocess here is probably reliable, but always makes me worry about details like signal propagation, passing through flags passed to the parent node process, etc

Reply:
This is certainly an option. There are a couple of reasons I went this way:
1. GitHub actions has first-class support for executing sub-processes (note that I'm using the GitHub library's `exec`), so it properly handles things like capturing output.
2. I want to force any new functionality added to the action to be supported by the CLI, so calling from the outer-most point enforces this.

Reply:
Including the code would allow us to stop vendoring two different bundles, as we're only bundling the CLI for this action.

Reply:
Keeping us honest that the CLI has feature-parity is a compelling reason, I'm fine with this either way.